src/merkle/mmr/partial.rs

```diff
@@ -173,6 +198,8 @@ impl PartialMmr {
         }
 
         let mut idx = InOrderIndex::from_leaf_pos(index);
+        self.nodes.insert(idx, node);
```

Comment:
In order to make `inner_nodes()` work, we need to keep track of the node itself (not only its sibling). I've modified only this method for now, but will also need to modify other relevant methods and tests. I believe the relevant methods are:

- `PartialMmr::apply()`.
- `PartialMmr::remove()`.

Another thing I'm wondering is whether keeping track of node values in the `nodes` vector would allow us to get rid of `track_latest` field (it seems like we should be able to figure out if we are tracking latest when the last node in `nodes` matches the last peak).

Reply:
> In order to make inner_nodes() work, we need to keep track of the node itself (not only its sibling).

This is very unfortunate, it will complicate this structure quite a bit. The presence of a node is used to signify the path is being tracked, conversely the lack of a node means the path is not tracked, here you're changing the meaning of `add` to track the paths to both this node and its sibling, as a consequence `remove` doesn't work anymore. This change will need a rewrite of the structure.

What is the use case driving this addition? Maybe we can implement it differently? The nodes must be tracked by another structure anyways, so one approach is to receive the list of tracked nodes as a parameter.

Reply:
The use case for this is for proving transaction execution where one of the inputs is `ChainMmr` struct (currently a simple wrapper around `Mmr`, but should be a wrapper around `PartialMmr`).

Yeah - I see how this complicates things. As you pointed out, there are two approaches:
1. Update the internal structure somehow to keep track of which leaves are tracked and which are not.
2. Modify `inner_nodes()` method to take a list of `(index, value)` tuples for the leaf nodes not in the `nodes` vector.

If there is a clean solution, the first approach would probably be better as it makes the data structure self-contained. On the other hand, the second approach could work fine for our purposes as we could put the extra info into `ChainMmr` struct.